### PR TITLE
Enable Extra Volumes on thanos-sidecar container

### PR DESCRIPTION
**Pull request by nicolastakashi**
Provide a capability to mount volumes on `thanos-sidecar` container, solving two existing issues: #3322 #3937

### CODE REVIEW

File: pkg/apis/monitoring/v1/types.go [hunk 1]
```diff
@@ -558,6 +558,12 @@ type ThanosSpec struct {
 	LogFormat string `json:"logFormat,omitempty"`
 	// MinTime for Thanos sidecar to be configured with. Option can be a constant time in RFC3339 format or time duration relative to current time, such as -1d or 2h45m. Valid duration units are ms, s, m, h, d, w, y.
 	MinTime string `json:"minTime,omitempty"`
+	// Volumes allows configuration of additional volumes on the output StatefulSet definition. Volumes specified will
+	// be appended to other volumes.
+	Volumes []v1.Volume `json:"volumes,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:563]: It seems redundant with the `Volumes` field at the spec level. Maybe we need only `VolumeMounts` here?

Comment (nicolastakashi) [reply pkg/apis/monitoring/v1/types.go:563]: I will take a look on that

Comment (nicolastakashi) [reply pkg/apis/monitoring/v1/types.go]: Hi @simonpasquier, sorry for the late reply.

May you take a look again?

Comment (nicolastakashi) [reply pkg/apis/monitoring/v1/types.go]: Hi @simonpasquier, sorry for the inconvenience.

Any news about that?

File: pkg/apis/monitoring/v1/types.go [hunk 2]
```diff
@@ -558,6 +558,9 @@ type ThanosSpec struct {
 	LogFormat string `json:"logFormat,omitempty"`
 	// MinTime for Thanos sidecar to be configured with. Option can be a constant time in RFC3339 format or time duration relative to current time, such as -1d or 2h45m. Valid duration units are ms, s, m, h, d, w, y.
 	MinTime string `json:"minTime,omitempty"`
+	// VolumeMounts allows configuration of additional VolumeMounts on the output StatefulSet definition.
+	// VolumeMounts specified will be appended to other VolumeMounts in the thanos-sidecar container,
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:562]:
```suggestion
	// VolumeMounts specified will be appended to other VolumeMounts in the thanos-sidecar container.
```

File: pkg/prometheus/statefulset_test.go [hunk 3]
```diff
@@ -1056,6 +1056,58 @@ func TestThanosTracing(t *testing.T) {
...
+			containsVolume = true
+			break
+		}
+	}
+	if !containsVolume {
+		t.Fatalf("Thanos sidecar volume is missing expected volume: %s", testVolume)
+	}
+
+	var containsVolumeMount bool
+	for _, volumeMount := range sset.Spec.Template.Spec.Containers[2].VolumeMounts {
```
Comment (simonpasquier) [inline pkg/prometheus/statefulset_test.go:1099]: (nit) I'd rather iterate over all `sset.Spec.Template.Spec.Containers` and check the one with name == "thanos-sidecar".

**Comment by nicolastakashi** [comment]
Sorry for the late answer!
Could you take a look again @simonpasquier?

**Review by paulfantom — APPROVED** [review]
lgtm, but leaving final say to @simonpasquier

**Comment by nicolastakashi** [comment]
Hi Guys!
@simonpasquier any news about that?

**Comment by nlamirault** [comment]
@nicolastakashi could you also add a `volume` field ?

To use Secret Store CSI Driver (like that https://github.com/GoogleCloudPlatform/secrets-store-csi-driver-provider-gcp/blob/main/examples/mypod.yaml.tmpl)

**Comment by simonpasquier** [comment]
@nlamirault the volumes are specified at the [PrometheusSpec](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/api.md#prometheusspec) level. My assumption is that we only have to expose volumeMounts for the thanos container. Did we miss anything?

@nicolastakashi could you rebase?

**Comment by nlamirault** [comment]
@simonpasquier good :+1:

**Comment by nicolastakashi** [comment]
@simonpasquier Sorry for the delay, can you take a look?

**Review by simonpasquier — APPROVED** [review]
lgtm

**Comment by simonpasquier** [comment]
@nicolastakashi can you run `make format` and commit?

**Comment by nicolastakashi** [comment]
@simonpasquier done dude, sorry for the delay, can you review it again?

**Comment by maikokuppe** [comment]
This would be highly appreciated because we also need this for using Thanos with TLS :)

**Comment by simonpasquier** [comment]
@nicolastakashi this would need yet another rebase on top of master. Sorry for the pain!

**Comment by maikokuppe** [comment]
This PR can be closed since these changes were merged in https://github.com/prometheus-operator/prometheus-operator/pull/4238.

**Pull request closed without merging**